apps/total-typescript/src/pages/api/skill/modules/verify-access.ts

```diff
@@ -0,0 +1,59 @@
+import {NextApiRequest, NextApiResponse} from 'next'
+import {getExercise} from 'lib/exercises'
+import {z} from 'zod'
+import {getToken} from 'next-auth/jwt'
+import {getSubscriberFromCookie} from 'server/ck-subscriber-from-cookie'
+import {getWorkshop} from 'lib/workshops'
+import {getTutorial} from 'lib/tutorials'
+import {getSection} from 'lib/sections'
+import {defineRulesForPurchases, UserSchema} from 'video/ability'
+
+const VerifyModuleAccessArgsSchema = z.object({
+  moduleSlug: z.string().optional(),
+  moduleType: z.string().optional(),
+  lessonSlug: z.string().optional(),
+  sectionSlug: z.string().optional(),
+  isSolution: z.boolean().default(false).optional(),
+})
+
+const verifyModuleAccess = async (
```

Comment:
same story here in terms of portability, but we've settled on the module/section/lesson structure and we could think about what that looks like in a portable sense more generically since we don't need the full details of a module to understand if the learner has access or not

in our current scheme for Total TypeScript we are giving folks access to the first lesson, so there is logic for that, but we might be able to be more explicit about that marking specific lessons as free to avoid the need for so many arguments and we strictly check if a learner has access to the module itself

with that in mind we can use the module id, convertkit subscriber, and the logged in user to determine current access levels like `can('view', 'Module', module)` and the rules contain a list of modules the learner has access to based on their combined `subscriber` + `user` state in the databases

Reply:
https://linear.app/skillrecordings/issue/TT-70/verify-access-at-the-module-level-instead-of-lesson-level